presto-orc/src/main/java/com/facebook/presto/orc/reader/NullWrappingColumnReader.java

```diff
@@ -0,0 +1,249 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.reader;
+
+import com.facebook.presto.orc.QualifyingSet;
+import com.facebook.presto.orc.stream.BooleanInputStream;
+import com.facebook.presto.orc.stream.LongInputStream;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+import static com.google.common.base.Verify.verify;
+
+abstract class NullWrappingColumnReader
+        extends ColumnReader
+        implements StreamReader
+{
+    QualifyingSet innerQualifyingSet;
+    boolean hasNulls;
+    int innerPosInRowGroup;
+    int numInnerRows;
+    int[] nullsToAdd;
+    int numNullsToAdd;
+    // Number of elements retrieved from inner reader.
+    int numInnerResults;
+
+    protected void beginScan(BooleanInputStream presentStream, LongInputStream lengthStream)
+            throws IOException
+    {
+        super.beginScan(presentStream, lengthStream);
+        numInnerResults = 0;
+    }
+
+    // Translates the positions of inputQualifyingSet between
+    // beginPosition and endPosition into an inner qualifying set for
+    // the non-null content.
+    protected void makeInnerQualifyingSets(int beginPosition, int endPosition)
+    {
+        if (presentStream == null) {
+            numInnerRows = endPosition - beginPosition;
+            hasNulls = false;
+            return;
+        }
+        hasNulls = true;
+        numInnerRows = 0;
+        if (innerQualifyingSet == null) {
+            innerQualifyingSet = new QualifyingSet();
+        }
+        QualifyingSet input = inputQualifyingSet;
+        int[] inputRows = input.getPositions();
+        int numActive = Math.min(endPosition, input.getPositionCount());
+        QualifyingSet inner = innerQualifyingSet;
+        int[] innerRows = inner.getMutablePositions(numActive);
+        int[] innerToOuter = inner.getMutableInputNumbers(numActive);
```

Comment:
innerToOuter name suggests a mapping, however, the type is an array. Also, the naming is inconsistent as innerToOutput is assigned to inputNumbers.

My understanding is that positions and inputNumbers are both array of ints and they have the same size. inputNumbers is used for simple types with complex encodings and complex types. For example, a list column has two streams: a stream of lengths and a string of individual elements (data). The size of stream of lengths is the number of non-null elements. The size of data stream is a sum of lengths. Let's say we have 3 rows of lists like so:

lengths = [3, 2, 1]
data = [10, 11, 12, 20, 21, 30]

Let's them assume that inputQualifyingSet contains two positions [0, 2].

Then, innerQualifyingSet should have positions = [0, 1, 2, 5] and inputNumbers = [0, 0, 0, 2].

Is this accurate? If so, I'd rename inputNumbers to outerPositions as the array contains outerPosition for each position in a nested data structure.